Action Errors preserves even if they are caught by programm flow

**Version**

2.4.1-next.405

**Short overview**

When you try and catch an error in the program flow of the sakuli check, the error is correctly caught so that the program flow resumes to the catch block, but the test will fail anyway with this error since it seems that sakuli "remembers" the error.

This behavior seems to happen everytime an action causes a error. 

**Detailed error description**

**Steps to reproduce error**

```typescript
(async () => {
    const tc = new TestCase("Issue");
    const env = new Environment();
    try {
        await _navigateTo('https://google.com/')

        try {
          await _click(_div('something-that-not-exists'))
        } catch(e) {
          Logger.logInfo('Well everything should be fine')
        }

    } catch (e) {
        await tc.handleException(e);
    } finally {
        await tc.saveResult();
    }
})()

```

In the example above the log Info `"Well everything should be fine"` is logged so, as a Testauthor I would assume that the test will run successfully.

Test log is as follows:

```
 ____            _              _   _ 
 / ___|    __ _  | | __  _   _  | | (_)
 \___ \   / _` | | |/ / | | | | | | | |
  ___) | | (_| | |   <  | |_| | | | | |
 |____/   \__,_| |_|\_\  \__,_| |_| |_|
                                       
========================================

Initialized Sakuli with 1 Testcases

Writing logs to: itsm/_logs/sakuli.log
[Tue Feb 02 2021 10:42:10 GMT+0000 (Coordinated Universal Time)] INFO: Loaded Sakuli with checkMkForwarderPreset,gearmanForwarderPreset,icinga2ForwarderPreset,prometheusPreset,legacyPreset,ocrPreset
[Tue Feb 02 2021 10:42:10 GMT+0000 (Coordinated Universal Time)] DEBUG: Checkmk Configuration:
enabled: false
spoolDir: /var/lib/check_mk_agent/spool
spoolfilePrefix: sakuli_suite
freshness: 600
piggybackHostName: 
sectionName: local
serviceDescription: 
outputDetails: true
url:
[Tue Feb 02 2021 10:42:10 GMT+0000 (Coordinated Universal Time)] DEBUG: Gearman Configuration:
enabled: false,
serverHost: 10.250.252.20,
serverPort: 4730,
serviceType: passive,
serverQueue: check_results,
encryption: true,
secretKey: ***,
nagiosHost: sakuli_client,
nagiosCheckCommand: check_sakuli,
nagiosServiceDescription: ,
nagiosTemplateSuiteSummaryMaxLength: 200,
nagiosTemplateScreenshotDivWidth: 640,
outputDetails: true
[Tue Feb 02 2021 10:42:10 GMT+0000 (Coordinated Universal Time)] DEBUG: Icinga2 Configuration:
enabled: false,
allowInsecure: false,
apiHost: undefined,
apiPort: 5665,
apiUserName: ,
apiPassword: ,
checkCommand: check_sakuli,
checkSource: check_sakuli,
hostName: ,
serviceDescription:
[Tue Feb 02 2021 10:42:10 GMT+0000 (Coordinated Universal Time)] DEBUG: Prometheus Forwarder Configuration:
enabled: false,
apiHost: ,
apiPort: 9091,
apiJob:
[Tue Feb 02 2021 10:42:10 GMT+0000 (Coordinated Universal Time)] INFO: Started Execution
[Tue Feb 02 2021 10:42:11 GMT+0000 (Coordinated Universal Time)] DEBUG: Created webdriver
[Tue Feb 02 2021 10:42:11 GMT+0000 (Coordinated Universal Time)] INFO: Started Testsuite itsm
[Tue Feb 02 2021 10:42:11 GMT+0000 (Coordinated Universal Time)] INFO: Started Testcase Issue
[Tue Feb 02 2021 10:42:11 GMT+0000 (Coordinated Universal Time)] DEBUG: New similarity value: 0.99
[Tue Feb 02 2021 10:42:12 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:13 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:13 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:13 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:13 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:13 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:13 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:13 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:14 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:14 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:14 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:14 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:14 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:14 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:14 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:14 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:15 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:15 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:15 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:15 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:15 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:15 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:15 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:15 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 120 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:15 GMT+0000 (Coordinated Universal Time)] DEBUG: Number of detected frames: 1
[Tue Feb 02 2021 10:42:15 GMT+0000 (Coordinated Universal Time)] DEBUG: Querying frame Nr.: 0
[Tue Feb 02 2021 10:42:16 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:16 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:16 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:16 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:16 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:16 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:16 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:16 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:17 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:17 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:17 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:17 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:17 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:17 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:17 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:17 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:18 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:18 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:18 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:18 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:18 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:18 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:18 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:18 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: Converted string "something-that-not-exists" to regex /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: Found 32 Elements for RegEx: /^\s*something-that-not-exists\s*$/
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: Received TimoutError when searching for element in frame Nr.: 0, not updating error history
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] INFO: Well everything should be fine
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] INFO: Finished Step UNNAMED with state Error 7.897s
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: Gearman forwarding disabled via properties.
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] INFO: Finished Testcase Issue with state Error 7.901s
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: Gearman forwarding disabled via properties.
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] INFO: Finished Testsuite itsm with state Error 8.001s
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: Gearman forwarding disabled via properties.
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: Closed webdriver
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] INFO: Finished Execution
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: CheckMK forwarding disabled via properties.
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: Gearman forwarding disabled via properties.
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: Icinga2 forwarding disabled via properties.
[Tue Feb 02 2021 10:42:19 GMT+0000 (Coordinated Universal Time)] DEBUG: Prometheus forwarding disabled via properties.

Failed to perform action click:

TimeoutError: Cannot find Element within 3000ms by query:
locator: By(css selector, div)
identifier: something-that-not-exists
Wait timed out after 3066ms
at /headless/.nvm/versions/node/v14.15.4/lib/node_modules/@sakuli/cli/node_modules/selenium-webdriver/lib/webdriver.js:842:17
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
npm ERR! Test failed.  See above for more details.
~
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -45,6 +45,7 @@ All notable changes to this project will be documented in this file.
 - Enhancement: Sakuli CLl encrypt command fails when secret is a numerical value [(#587)](https://github.com/sakuli/sakuli/issues/587)
 - Enhancement: Add stack traces to forwader error logging [(#601)](https://github.com/sakuli/sakuli/issues/601)
 - Bugfix: throwing/rejecting strings instead of errors causing issues in error handling [(#600)](https://github.com/sakuli/sakuli/issues/600)
+- Enhancement: Action Errors preserves even if they are caught by programm flow [(#597)](https://github.com/sakuli/sakuli/issues/597)
 
 ## v2.4.0
 

--- a/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.spec.ts
+++ b/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.spec.ts
@@ -8,6 +8,7 @@ describe("create-with-action-context", () => {
     name: string,
     fn: (...args: ARGS) => Promise<R>
   ) => (...args: ARGS) => Promise<R>;
+  const actionName = "dummy";
 
   beforeEach(() => {
     ctx = createTestExecutionContextMock();
@@ -20,15 +21,18 @@ describe("create-with-action-context", () => {
 
     beforeEach(() => {
       actionMock = jest.fn().mockResolvedValue("Works");
-      wrappedAction = withActionContext("dummy", actionMock);
+      wrappedAction = withActionContext(actionName, actionMock);
     });
 
     it("should invoke action and return its result", async () => {
       await wrappedAction(1, "Test");
       expect(ctx.startTestAction).toHaveBeenCalledWith(
-        expect.objectContaining({ id: "dummy" })
+        expect.objectContaining({ id: actionName })
       );
-      expect(ctx.logger.trace).toHaveBeenCalledTimes(2);
+      expect(ctx.logger.trace).toHaveBeenNthCalledWith(
+        1,
+        `Start action ${actionName}`
+      );
       expect(ctx.endTestAction).toHaveBeenCalled();
     });
   });
@@ -41,27 +45,53 @@ describe("create-with-action-context", () => {
     beforeEach(() => {
       actionError = Error("Error in Action");
       actionMock = jest.fn().mockRejectedValue(actionError);
-      wrappedAction = withActionContext("dummy", actionMock);
+      wrappedAction = withActionContext(actionName, actionMock);
     });
 
     it("should invoke action and throw an error", async () => {
+      //GIVEN
+      const error = Error("Error in Action");
+
+      //WHEN
       const wrappedActionResultPromise = wrappedAction(1, "Test");
-      await expect(wrappedActionResultPromise).rejects.toThrowError(
-        /Error in Action/
-      );
+
+      //THEN
+      await expect(wrappedActionResultPromise).rejects.toThrowError(error);
       expect(actionMock).toHaveBeenCalledWith(1, "Test");
-      expect(ctx.logger.trace).toHaveBeenCalledTimes(2);
+      expect(ctx.logger.trace).toHaveBeenNthCalledWith(
+        1,
+        `Start action ${actionName}`
+      );
+      expect(ctx.logger.trace).toHaveBeenNthCalledWith(
+        2,
+        `Finish action ${actionName}`,
+        expect.stringMatching(error.message)
+      );
+      expect(ctx.getCurrentTestAction).toHaveBeenCalledTimes(1);
     });
 
     it("should start and finish an action anyway", async () => {
-      await expect(wrappedAction(1, "Test")).rejects.toThrowError(
-        /Error in Action/
-      );
+      //GIVEN
+      const error = Error("Error in Action");
+
+      //WHEN
+      await expect(wrappedAction(1, "Test"))
+        //THEN
+        .rejects.toThrowError(error);
       expect(ctx.startTestAction).toHaveBeenCalledWith(
         expect.objectContaining({ id: "dummy" })
       );
       expect(ctx.endTestAction).toHaveBeenCalled();
-      expect(ctx.logger.trace).toHaveBeenCalledTimes(2);
+      expect(ctx.logger.trace).toHaveBeenNthCalledWith(
+        1,
+        `Start action ${actionName}`
+      );
+      expect(ctx.logger.trace).toHaveBeenNthCalledWith(
+        2,
+        `Finish action ${actionName}`,
+        expect.stringMatching(error.message)
+      );
+      expect(ctx.getCurrentTestAction).toHaveBeenCalledTimes(1);
     });
   });
 });

--- a/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.spec.ts
+++ b/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.spec.ts
@@ -64,7 +64,7 @@ describe("create-with-action-context", () => {
       );
       expect(ctx.logger.trace).toHaveBeenNthCalledWith(
         2,
-        `Finish action ${actionName}`,
+        `Finish action ${actionName} with errors`,
         expect.stringMatching(error.message)
       );
       expect(ctx.getCurrentTestAction).toHaveBeenCalledTimes(1);
@@ -88,7 +88,7 @@ describe("create-with-action-context", () => {
       );
       expect(ctx.logger.trace).toHaveBeenNthCalledWith(
         2,
-        `Finish action ${actionName}`,
+        `Finish action ${actionName} with errors`,
         expect.stringMatching(error.message)
       );
       expect(ctx.getCurrentTestAction).toHaveBeenCalledTimes(1);

--- a/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.spec.ts
+++ b/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.spec.ts
@@ -1,6 +1,6 @@
 import { createWithActionContext } from "./create-with-action-context.function";
 import { createTestExecutionContextMock } from "../../../__mocks__";
-import { TestExecutionContext } from "@sakuli/core/dist";
+import { TestExecutionContext } from "@sakuli/core";
 
 describe("create-with-action-context", () => {
   let ctx: TestExecutionContext;
@@ -59,9 +59,6 @@ describe("create-with-action-context", () => {
       );
       expect(ctx.startTestAction).toHaveBeenCalledWith(
         expect.objectContaining({ id: "dummy" })
-      );
-      expect(ctx.updateCurrentTestAction).toHaveBeenCalledWith(
-        expect.objectContaining({ error: actionError })
       );
       expect(ctx.endTestAction).toHaveBeenCalled();
       expect(ctx.logger.trace).toHaveBeenCalledTimes(2);

--- a/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.ts
+++ b/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.ts
@@ -1,30 +1,39 @@
 import { TestExecutionContext } from "@sakuli/core";
 import { ifPresent } from "@sakuli/commons";
+import * as util from "util";
 
 export function createWithActionContext(ctx: TestExecutionContext) {
   return function runAsAction<ARGS extends any[], R>(
     name: string,
     fn: (...args: ARGS) => Promise<R>
   ): (...args: ARGS) => Promise<R> {
+    function logFinishedAction(error: any) {
+      const log = [`Finish action ${name}`];
+      ifPresent(ctx.getCurrentTestAction(), (action) => {
+        log.push(error ? `with errors` : `successfully`);
+        log.push(
+          `after ${
+            (new Date().getTime() - action.startDate!.getTime()) / 1000
+          }s`
+        );
+      });
+
+      util.types.isNativeError(error)
+        ? ctx.logger.trace(log.join(" "), error.stack)
+        : ctx.logger.trace(log.join(" "));
+    }
     return async (...args: ARGS): Promise<R> => {
       ctx.startTestAction({ id: name });
       ctx.logger.trace(`Start action ${name}`);
       let res: any;
+      let actionError = undefined;
       try {
         res = await fn(...args);
       } catch (error) {
+        actionError = error;
         throw error;
       } finally {
-        const log = [`Finish action ${name}`];
-        ifPresent(ctx.getCurrentTestAction(), (action) => {
-          log.push(action.error ? `with errors` : `successfully`);
-          log.push(
-            `after ${
-              (new Date().getTime() - action.startDate!.getTime()) / 1000
-            }s`
-          );
-        });
-        ctx.logger.trace(log.join(" "));
+        logFinishedAction(actionError);
         ctx.endTestAction();
       }
       return res;

--- a/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.ts
+++ b/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.ts
@@ -13,7 +13,6 @@ export function createWithActionContext(ctx: TestExecutionContext) {
       try {
         res = await fn(...args);
       } catch (error) {
-        ctx.updateCurrentTestAction({ error });
         throw error;
       } finally {
         const log = [`Finish action ${name}`];
@@ -21,7 +20,7 @@ export function createWithActionContext(ctx: TestExecutionContext) {
           log.push(action.error ? `with errors` : `successfully`);
           log.push(
             `after ${
-              (new Date().getTime() - action!.startDate!.getTime()) / 1000
+              (new Date().getTime() - action.startDate!.getTime()) / 1000
             }s`
           );
         });

--- a/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.ts
+++ b/packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.ts
@@ -1,39 +1,33 @@
 import { TestExecutionContext } from "@sakuli/core";
-import { ifPresent } from "@sakuli/commons";
-import * as util from "util";
 
 export function createWithActionContext(ctx: TestExecutionContext) {
   return function runAsAction<ARGS extends any[], R>(
     name: string,
     fn: (...args: ARGS) => Promise<R>
   ): (...args: ARGS) => Promise<R> {
-    function logFinishedAction(error: any) {
-      const log = [`Finish action ${name}`];
-      ifPresent(ctx.getCurrentTestAction(), (action) => {
-        log.push(error ? `with errors` : `successfully`);
-        log.push(
-          `after ${
-            (new Date().getTime() - action.startDate!.getTime()) / 1000
-          }s`
-        );
-      });
-
-      util.types.isNativeError(error)
-        ? ctx.logger.trace(log.join(" "), error.stack)
-        : ctx.logger.trace(log.join(" "));
+    function createActionDurationString() {
+      const actionStartTime = ctx.getCurrentTestAction()?.startDate?.getTime();
+      return actionStartTime
+        ? ` after ${(new Date().getTime() - actionStartTime) / 1000}s`
+        : "";
     }
+
     return async (...args: ARGS): Promise<R> => {
       ctx.startTestAction({ id: name });
       ctx.logger.trace(`Start action ${name}`);
       let res: any;
-      let actionError = undefined;
       try {
         res = await fn(...args);
+        ctx.logger.trace(
+          `Finish action ${name} successfully${createActionDurationString()}`
+        );
       } catch (error) {
-        actionError = error;
+        ctx.logger.trace(
+          `Finish action ${name} with errors${createActionDurationString()}`,
+          error.stack
+        );
         throw error;
       } finally {
-        logFinishedAction(actionError);
         ctx.endTestAction();
       }
       return res;

